crates/iota-data-ingestion/src/chain_tip_tracker.rs

```diff
@@ -0,0 +1,119 @@
+                ..
+            } => Some(*first_chk_seq_num_of_epoch),
+            Self::CurrentEpoch { .. } => None,
+        }
+    }
+}
+
+/// Tracks the current state of checkpoints at the tip of the network.
+pub struct ChainTipWatermarkTracker {
+    client: Client,
+    last_worker_watermark: Option<CheckpointSequenceNumber>,
+}
+
+impl ChainTipWatermarkTracker {
+    /// Creates a new chain tip tracker.
+    pub fn new(
+        node_rest_api_url: &str,
+        last_worker_watermark: Option<CheckpointSequenceNumber>,
+    ) -> Self {
+        Self {
+            client: Client::new(node_rest_api_url),
+            last_worker_watermark,
+        }
+    }
+
+    /// Resolves the current state at the network tip.
+    pub async fn resolve_network_tip_state(&self) -> anyhow::Result<NetworkTipState> {
+        let latest_checkpoint = self.client.get_latest_checkpoint().await?;
+
+        let latest_watermark_checkpoint: OptionFuture<_> = self
+            .last_worker_watermark
+            .map(|chk_seq_num| self.client.get_checkpoint_summary(chk_seq_num))
+            .into();
+
+        let status = match latest_watermark_checkpoint.await {
+            Some(Ok(summary)) if summary.epoch != latest_checkpoint.epoch => {
+                let watermark = self
+                    .find_first_checkpoint_of_current_epoch(&latest_checkpoint)
+                    .await?;
+                NetworkTipState::EpochChanged {
+                    epoch: latest_checkpoint.epoch,
+                    first_chk_seq_num_of_epoch: watermark,
+                }
+            }
+            Some(Ok(summary)) => NetworkTipState::CurrentEpoch {
+                epoch: summary.epoch,
+            },
+            _ => {
+                let watermark = self
+                    .find_first_checkpoint_of_current_epoch(&latest_checkpoint)
+                    .await?;
+                NetworkTipState::EpochChanged {
+                    epoch: latest_checkpoint.epoch,
+                    first_chk_seq_num_of_epoch: watermark,
+                }
+            }
+        };
+
+        Ok(status)
+    }
+
+    /// Finds the first checkpoint sequence number of a given epoch.
+    async fn find_first_checkpoint_of_current_epoch(
+        &self,
+        latest_checkpoint: &CertifiedCheckpointSummary,
+    ) -> anyhow::Result<CheckpointSequenceNumber> {
+        if latest_checkpoint.sequence_number == 0 || latest_checkpoint.epoch == 0 {
+            return Ok(0);
+        }
+
+        let previous_epoch = latest_checkpoint.epoch.saturating_sub(1);
+
+        let epoch_last_checkpoint = self
+            .client
+            .get_epoch_last_checkpoint(previous_epoch)
+            .await?;
+        let target_epoch_first_checkpoint = epoch_last_checkpoint.sequence_number + 1;
+
+        tracing::info!(
+            "Found first checkpoint of epoch {}: checkpoint {target_epoch_first_checkpoint}",
+            latest_checkpoint.epoch
+        );
```

Comment:
Should this really be ::info! ? looks more like it should be ::debug! to me, but feel free to resolve as I don't know how exactly this will be used

Reply:
removed that log as part of the simplification of the overall logic: https://github.com/iotaledger/iota/pull/5978/commits/a926eaf991ede958c3c2d12e3c1bd2e784e7bda1